### PR TITLE
Quickbuild mpif08

### LINKED ISSUE #261: mpich mpif08
I think the mpi_utilities errors are mpich using f77 bindings when you `use mpi`. Hello world mpi (no dart code) will not compile if you have a mpi_send(integer) mpi_send(real) in the same file.

[hello.f90.txt](https://github.com/NCAR/DART/files/6525415/hello.f90.txt)
[hello_f08.f90.txt](https://github.com/NCAR/DART/files/6525419/hello_f08.f90.txt)

The f08 mpi module `use mpi_f08` follows the fortran standard. 
There is a branch mpi_f08 which has a version of mpi_utilites_mod.f90 with mpi_f08 bindings. This will work with mpich and gfortran 10+ (only done a quick test).
The cray window module (cray pointers) could get replaced with iso_c_bindings. 

or short term solution: `port select --set mpi openmpi-mp-fortran`

_Originally posted by @hkershaw-brown in https://github.com/NCAR/DART/issues/158#issuecomment-846307106_

### LINKED ISSUE #497: Documentation needs more information about the "quickbuild nompi" option
Our SIParCS students ran into the same issue when building and running lorenz_63 with DART for the first time. This was that they tried to run filter after running "quickbuild.sh" without having an MPI library available on their local machines, so it failed. 

The way to fix this is to simply add "nompi" to the quickbuild command, but this was not apparent to our students working off our guide to compiling DART ([https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html](https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html)), which makes no mention of it. 

While the option is discussed in the "quick-start" section of the home page ([https://docs.dart.ucar.edu/en/latest/README.html#quick-start] 
(https://docs.dart.ucar.edu/en/latest/README.html#quick-start)), it can easily be missed here. 

edit README.html is now index.html
https://docs.dart.ucar.edu/en/latest/index.html#quick-start

This will be a really simple documentation fix that will add some basic information on how to use the "quickbuild nompi" option to the pages that are needing it:
[https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html](https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html)
[https://docs.dart.ucar.edu/en/latest/guide/quickbuild.html#dart-build-system](https://docs.dart.ucar.edu/en/latest/guide/quickbuild.html#dart-build-system)

This will just make this option a little more obvious to our users, especially those building DART for the first time. 
 
Using DART v10.7.3

### LINKED ISSUE #551: bug: quickbuild.sh errors out with the gnu compiler on Derecho (MPI errors)
### Describe the bug

1. List the steps someone needs to take to reproduce the bug.  
Run quickbuild.sh on any model on Derecho using the gnu compiler
 
2. What was the expected outcome?
Filter finishes successfully

3. What actually happened?  
Multiple MPI error messages (rank and type mismatches) are produced from mpi_utilities_mod.f90

### Error Message  
See screenshots

### Which model(s) are you working with?
All models

### Screenshots   
image
image

### Version of DART      
Which version of DART are you using? 
v10.8.4

### Have you modified the DART code?   
No

### Build information
Please describe:  
 1. The machine you are running on (e.g. windows laptop, NCAR supercomputer Cheyenne).   
 2. The compiler you are using (e.g. gnu, intel).  

Derecho, gnu

### PR DESCRIPTION
**Pull request by hkershaw-brown**
## Description:

Added an mpif08 option to quickbuild.sh for model quickbuilds.

I've added mpi as an option for the 1st argument to quickbuild.sh.  This is because it seemed likely that someone might specify `./quickbuild.sh mpi` if they can do `./quickbuild.sh mpif08`

The default is still mpi.

The GSI converter uses mpi, but its own mpi calls (not mpif08)
https://github.com/NCAR/DART/blob/main/observations/obs_converters/GSI2DART/mpisetup.f90
I have left this alone for now.

### Fixes issue

Fixes #261, fixes #551, fixes #497 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?

- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
   added the fix for #497 in this pull request.

### Tests
Building, with mpi, mpif08, nompi
`./quickbuild.sh help`  To print usage message.
will test on Derecho with the various compilers and mpi libraries once derecho is a back up.

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

### CODE REVIEW

File: build_templates/buildfunctions.sh [hunk 1]
```diff
@@ -146,24 +163,39 @@ local misc="$DART/models/utilities/ \
...
 local nullwin="$DART"/assimilation_code/modules/utilities/null_win_mod.f90
 local craywin="$DART"/assimilation_code/modules/utilities/cray_win_mod.f90
 local nocraywin="$DART"/assimilation_code/modules/utilities/no_cray_win_mod.f90
+local no_cray_winf08="$DART"/assimilation_code/modules/utilities/no_cray_winf08_mod.f90
 
 if [ "$mpisrc" == "mpi" ]; then
 
    core=${core//$nullmpi/}
    core=${core//$nullwin/}
+   core=${core//$mpif08/}
```
Comment (nancycollins) [inline build_templates/buildfunctions.sh:177]: this looks backwards to me - is it just the labels?  mpi includes the nullmpi version, and below the \#nompi version includes the mpi modules?

**Review by nancycollins — COMMENTED** [review]
everything looks fine except i was confused by the "core" lists in the buildfunctions.sh for the mpi vs nompi cases.  it's probably how it should be but it looks strange on first read.

File: build_templates/buildfunctions.sh (hunk 1 above)
Comment (hkershaw-brown) [reply build_templates/buildfunctions.sh:177]: It does look backwards with the `//`  at the beginning, but if you believe me:

`core=${core//$nullmpi/}` 

replace  all matches (`//` ) of`$nullmpi` from `$core` with nothing `('/ ')`

so for mpi, remove null, nullwin, mpif08, mpif08, no_cray_winf08

**Review by nancycollins — APPROVED** [review]
looks good.

**Comment by hkershaw-brown** [comment]
needs converters

**Review by mjs2369 — COMMENTED** [review]
https://github.com/NCAR/DART/blob/6c6ac535ff2a75d2b1831b5e414570cf370a3a66/build_templates/buildconvfunctions.sh#L45C1-L55C2

In the function print_usage in buildconvfunctions.sh, it should echo ./quickbuild.sh instead of buildconverter.sh

Also, is line 52 (copied below) true? I don't think that we can build individual programs with obs_converters
`echo "  buildconverter.sh [program]     : build a single program"`

Attempting to run `./quickbuild.sh filter` while in the work dir for an obs converter results in "ERROR: unknown program  filter"

**Comment by hkershaw-brown** [comment]
> https://github.com/NCAR/DART/blob/6c6ac535ff2a75d2b1831b5e414570cf370a3a66/build_templates/buildconvfunctions.sh#L45C1-L55C2
> 
> In the function print_usage in buildconvfunctions.sh, it should echo ./quickbuild.sh instead of buildconverter.sh
> 
> Also, is line 52 (copied below) true? I don't think that we can build individual programs with obs_converters `echo " buildconverter.sh [program] : build a single program"`
> 
> Attempting to run `./quickbuild.sh filter` while in the work dir for an obs converter results in "ERROR: unknown program filter"

Fixed  buildconverter.sh -> quickbuild.sh

The converter quickbuilds will build a single program, but it has to be one of the converter programs.  Do you think it would be less confusing to just have the docs say run ./quickbuild.sh help in a work directory to see the options?

**Comment by mjs2369** [comment]
>>
> The converter quickbuilds will build a single program, but it has to be one of the converter programs. Do you think it would be less confusing to just have the docs say run ./quickbuild.sh help in a work directory to see the options?

I think that is already covered in the docs, unless you mean that running "./quickbuild help" will show the options for possible single programs that can be built for the given work dir, which it currently doesn't do

**Comment by hkershaw-brown** [comment]
> > > 
> > 
> > 
> > The converter quickbuilds will build a single program, but it has to be one of the converter programs. Do you think it would be less confusing to just have the docs say run ./quickbuild.sh help in a work directory to see the options?
> 
> I think that is already covered in the docs, unless you mean that running "./quickbuild help" will show the options for possible single programs that can be built for the given work dir, which it currently doesn't do

To address https://github.com/NCAR/DART/issues/497,   I put in the the usage for quickbuild.sh for lorenz_63. 
but it seems like this is confusing for converter builds.  What were you thinking for #497 ?

**Comment by mjs2369** [comment]
>>
> To address #497, I put in the the usage for quickbuild.sh for lorenz_63.
> but it seems like this is confusing for converter builds. What were you thinking for #497 ?

I really like the additions to the usage for quickbuild.sh for lorenz_63 and think it fixes 497 well. I don't think it's confusing for the obs converters; if they try to use the mpi options it will fail and from there they can use the help command. 

But we could maybe add something like "not for obs converters" to this line [here](https://dart-documentation--562.org.readthedocs.build/en/562/guide/compiling-dart.html#:~:text=quickbuild.sh%20%5Bmpi/nompi/mpif08%5D%20%5Bprogram%5D%20%20%20%3A%20optional%20arguments)

**Review by mjs2369 — APPROVED** [review]
I think the compiling-dart doc is great! Ready to merge

**Pull request merged**